yamux/README.md

```diff
@@ -152,8 +152,18 @@ The Length should be set to one of the following to provide an error code:
 
 ## Implementation considerations
 
-### ACK backlog
+### ACK backlog & backpressure
 
 Yamux allows for a stream to be opened (and used) before it is acknowledged by the remote.
+Yamux also does not specify a backpressure mechanism for opening new streams.
+
+This presents a problem:
+A peer must read from the socket and decode the frames to make progress on existing streams.
+But any frame could also open yet another stream.
+
 The ACK backlog is defined as the number of streams that a peer has opened which have not yet been acknowledged.
-Implementations SHOULD at most allow an ACK backlog of 256 streams.
+To support a basic form of backpressure, implementions:
+
+- SHOULD at most allow an ACK backlog of 256 streams.
+- SHOULD buffer at least 256 unacknowledged inbound streams.
```

Comment:
Why? Resetting is a totally valid thing to do if you don’t want to buffer 256.

Reply:
The practical issue is that the first thing that happens on a stream is multistream-select. Resetting a stream will be treated by most existing implementations as a negotiation failure.

So whilst the yamux spec allows this, existing implementations and our usage of yamux in libp2p doesn't really cope very well with that.

If I understand correctly, the meaning of SHOULD is that it is completely optional but when supported, gives access to more features.

In this case, buffering 256 streams allows an implementation to trigger the 256-backlog backpressure mechanism also documented in this section.

If you reset a stream, the other peer will just continue to open new streams. Interpreting _that_ as a way of signaling backpressure was actually my original idea in #471 but we moved away from that in favor of following go-libp2p's implementation.

Reply:
Friendly ping @marten-seemann: Does the above explanation make sense? I'd like to close off this particular work stream so I'd appreciate some concrete suggestions on what to adjust in this PR so we can merge it and accurately reflect the status quo.

Reply:
Yes, my objections still stand.

Reply:
Can we merge it with we change it to `MAY`?